---
title: "Untitled"
output: html_document
date: "2023-10-06"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
library(dplyr)
library(ggplot2)
library("readxl")
library(gridExtra)
library(cowplot)
```

```{r}
setwd("C:/Users/deani/Desktop/Specialeprojekt")
getwd()
```
```{r}
ndf100 <- read.table("ndf100.tsv", header = TRUE, sep = "\t")
df <- ndf100 %>% select(strain, file, domain, phylum, class, order, family, genus, species, Plate, ONT_bc, DNA_conc, filtered_data, genomesize, Contigs, coverage, N50_assembly)
df
```

## Adding sequencing and assembly quality info

# Reading table and assigning meaningful column names
```{r}
info <- read_excel("Overview_of_all_isolates_2022.xlsx")
info <- info %>% rename(genome_size=`Genome size (Mb)`, strain=Strain, N50_reads=`N50 reads`, coverage=`Avg cov`, ONT_bc=`ONT bc`, DNA_conc=`DNA conc. used in lib (ng/Âµl)`, read_counts=`Read counts`, filtered_data=`Filtered_data (Mb)`, median_reads = `Median reads (bp)`, mean_reads = `Mean reads (bp)`, read_len_stdev = `Read Len. stdev`, N50_assembly=`N50 assembly (bp)`, lineage=`Lineage (tentative)`) %>% filter(!is.na(genome_size))
#head(info)
```

# Combining tables
```{r}
ndf100 <- left_join(ndf100,info, by="strain")
#ndf100
```

# Adjusting the genome size to bp from Mb
```{r}
ndf100 <- ndf100 %>% mutate(genomesize = case_when(genome_size < 15 ~ genome_size*1000000, genome_size > 15 ~ genome_size))
#ndf100$genome_size
#ndf100$genomesize
```

# Descriptive statistics of assembly quality markers
```{r}
summary(ndf100$genomesize)
quantile(ndf100$genomesize, 0.1)
quantile(ndf100$genomesize, 0.9)
summary(ndf100$Contigs)
summary(ndf100$coverage)
summary(ndf100$N50_assembly)
```
With plots 

```{r}
rects1 <- data.frame(xstart = c(1500000,4060099,7543137), xend = c(4060099, 7543137, 15000000), col = c("a", "b", "a"))
rects2 <- data.frame(xstart = c(-1,7), xend = c(7, 100), col = c("b", "a"))
rects3 <- data.frame(xstart = c(0,13), xend = c(13, 250), col = c("a", "b"))
rects4 <- data.frame(xstart = c(0,3470091), xend = c(3470091, 6500000), col = c("a", "b"))

pl1 <- ggplot() +
  geom_rect(data = rects1, aes(xmin = xstart, xmax = xend, ymin = -Inf, ymax = Inf, fill = col), alpha = 0.4) +
  geom_histogram(data = ndf100, aes(x=genomesize), binwidth=250000, color="black", fill="white") +
  geom_vline(aes(xintercept =4060099), color="black", linetype="dashed", size=0.5) +
  geom_vline(aes(xintercept =7543137), color="black", linetype="dashed", size=0.5) +
  xlim(1500000,15000000)

pl2 <- ggplot() +
  geom_rect(data = rects2, aes(xmin = xstart, xmax = xend, ymin = -Inf, ymax = Inf, fill = col), alpha = 0.4) +
  geom_histogram(data = ndf100, aes(x=Contigs), binwidth=2, color="black", fill="white") +
  geom_vline(aes(xintercept =7.00), color="black", linetype="dashed", size=0.5) +
  xlim(-1,100)

pl3 <- ggplot() +
  geom_rect(data = rects3, aes(xmin = xstart, xmax = xend, ymin = -Inf, ymax = Inf, fill = col), alpha = 0.4) +
  geom_histogram(data = ndf100, aes(x=coverage), binwidth = 5, color="black", fill="white") +
  geom_vline(aes(xintercept =13.00), color="black", linetype="dashed", size=0.5)

pl4 <- ggplot() + 
  geom_rect(data = rects4, aes(xmin = xstart, xmax = xend, ymin = -Inf, ymax = Inf, fill = col), alpha = 0.4) +
  geom_histogram(data=ndf100, aes(x=N50_assembly), binwidth=100000, color="black", fill="white") +
  geom_vline(aes(xintercept =3470091), color="black", linetype="dashed", size=0.5)

plot_grid(pl1, pl2, pl3, pl4, labels=c("A", "B", "C", "D"), ncol = 2, nrow = 2)
```






## Filtering the genomes with 100% identity using assembly quality

```{r}
filtered <- ndf100 %>% filter(Contigs <= 7) %>% filter(coverage >= 13) %>% filter(N50_assembly >= 3470091) %>% filter(genomesize >= 4060099) %>% filter(genomesize <= 7543137) 

filtered

#filtered <- ndf100 %>% filter(Contigs <= 10) %>% filter(coverage > 20) %>% filter(N50_assembly > 50000) %>% arrange(desc(coverage))
#filtered

```
# Writing the resulting list of genomes to a file

```{r}
#filtered_list <- filtered %>% mutate(file = paste0(strain, ".fasta")) %>% select(file)
#write.table(filtered_list, file = "filtered.txt", sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)

filtered_list <- filtered %>% mutate(file = paste0(strain, ".fasta")) %>% select(file)
filtered_list
write.table(filtered_list, file = "filtered.txt", sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

```{r}
filtered_list3 <- filtered %>% mutate(file = paste0(strain, ".phages_combined.fna")) %>% select(file)

#write.table(filtered_list3, file = "filtered_phages.txt", sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)
```


## Adding information of lengths of individual contigs (the .fasta.fai files combined)

```{r}
index <- read.table("index.tsv", header = FALSE) # read the file

n <- c("contig", "contig_nr", "strain") # columns to separate contig info into
index <- index %>% separate(V1, into = n, sep = "_") %>% rename(contig_length = V2) %>% select(contig_nr, strain, contig_length)
#head(index)
```

```{r}
# sort after contig length and keep one contig per strain: creates a list of the longest contig from each assembly
index <- index %>% arrange(desc(contig_length)) %>% distinct(strain, .keep_all = TRUE) %>% rename(longest_contig = contig_length)
#head(index)
```


```{r}
ndf100 <- left_join(ndf100,index, by="strain")
```

## Filtering for minimum length of the largest contig
The largest contig should be genome-sized (doesn't remove any rows after the previous filtering)

```{r}
dim(ndf100 %>% filter(Contigs <= 10) %>% filter(coverage > 20) %>% filter(N50_assembly > 50000) %>% filter(longest_contig>3000000))
```

```{r}
strain_list <- as.tibble(read.table("complete_strains.txt"))
conversion <- as.tibble(read_tsv("barcode_strain_conversion.tsv", col_names = FALSE))

#strain_list
#conversion
```

```{r}
strain_list <- strain_list %>% mutate(file = str_extract(V1, pattern = "Plate.*.txt"))
strain_list <- strain_list %>% select(file)
strain_list
```

```{r}
conversion <- conversion %>% rename(strain=X1, file=X2)
conversion
```


```{r}
complete_strains <- left_join(strain_list, conversion, by="file") %>% select(strain) %>% mutate(complete="yes")
complete_strains
```

```{r}
df100 <- left_join(df100, complete_strains, by="strain")
df100
```
```{r}
plate8 <- df100 %>% filter(Plate == "Plate 8")
rest <- df100 %>% filter(Plate != "Plate 8")

plate8 %>% filter(complete == "yes")
rest %>% filter(complete == "yes")
```

```{r}
new_file_list <- rest %>% filter(complete == "yes") %>% select(file)
new_file_list1 <- plate8 %>% filter(Contigs < 8) %>% filter(coverage > 20) %>% select(file)

new_file_list
new_file_list1

filtered <- bind_rows(new_file_list, new_file_list1)
filtered
```
```{r}
write.table(new_file_list, file = "new_list.txt", sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(new_file_list1, file = "new_list1.txt", sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

```{r}
disc <- read_excel("ani_99_discard.xlsx", col_names = FALSE)
disc
```
```{r}
discard <- disc %>% rename(file = "...1") %>% mutate(discard = "yes")
discard
```

```{r}
molly <- left_join(filtered,discard, by="file")
molly1 <- molly %>% filter(is.na(discard)) %>% select(file)
```

```{r}
write.table(molly1, file = "selected_strains.txt", sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

